PCF-ADR-0005-mcad-api-groups-migration.md

```diff
@@ -0,0 +1,131 @@
+| Superseded by: | N/A                      |
+| Issues         | https://github.com/project-codeflare/multi-cluster-app-dispatcher/issues/352 |
+
+## What
+
+The MCAD APIs are currently defined in the `mcad.ibm.com` and `ibm.com` groups.
+The objective of this document is to propose a plan to migrate them to the `.codeflare.dev` top-level group.
+
+This migration will also be the opportunity to organize the APIs by functional areas, i.e., the `workload.codeflare.dev` group for workload management APIs, and the `quota.codeflare.dev` group for quota management APIs.
+
+## Why
+
+It's conventional for open-source projects to use their domain name as their top-level API group, so DNS is leveraged as a global naming system.
+
+The choice of the current MCAD API groups predates the acquisition of the `codeflare.dev` domain, which explains why they do not follow that convention.
+
+Some of the MCAD APIs, mainly the AppWrapper API, is used by downstream projects, as well as end-users.
+From a client standpoint, the impacts of an API group change imply updating the `apiVersion` field of all the MCAD resources they define, as well as updating the group segment of the REST endpoints, created from the CRDs, that they consume.
+
+## Goals
+
+The goal is to inform a migration path, to provide these clients with a transition to the new API groups, that balances compatibility and costs.
+
+## Non-Goals
+
+It is not a goal to provide an automated migration.
+Downstream projects and end-users will have to address the impacts manually to perform the migration to the new API groups.
+
+## How
+
+A project like [TorchX](https://github.com/pytorch/torchx), which integrates with MCAD[^1] via the AppWrapper API, has its own release cadence, and it'd be challenging to have it migrated within a single migration cycle, without breaking compatibility.
+
+[^1]: https://pytorch.org/torchx/latest/schedulers/kubernetes_mcad.html
+
+The following plan proposes a progressive migration path, implemented in three phases, as follows:
+
+### Phase 1: Add Dual API Groups Support
+
+The new API groups are introduced in MCAD, along with the existing ones.
+Functionally, the MCAD and InstaScale controllers are updated, so they can equally reconcile the old and new APIs.
+
+To implement that dual API groups support, it may be necessary to duplicate the MCAD API packages, as well as the generated client packages.
+Indeed, [controller-runtime](https://github.com/kubernetes-sigs/controller-runtime) APIs, which are used by InstaScale, are designed to use the API structs, and require the mapping between the API structs and the Group/Version/Kind(GVK)s to be an injection.
+Otherwise stated, controller-runtime mandates the scheme holding the mapping between the API structs and the GVKs to have a single GVK entry per API struct.
+
+This boils down to the following tasks:
+
+* Duplicate the API packages, with the new API groups, and mark the old ones as deprecated
```

Comment:
I think that would be good for this document that the package names / structures are presented in this ADR.

Also, having support in MCAD for two api groups might be a lot of work and make the existing codebase even more incomprehensible. Is it worth adding a "conversion" web hook so that the changes to MCAD codebase would be more 'sane' ? 

If yes, does it make sense to explicitly mention it here ?

Reply:
Conversion webhooks address the conversion across versions of the same CRD. The `.spec.group` field is immutable, as the other fields that uniquely identify the CRD, e.g., the name, which implies the change of API group is actually the creation of an entirely new CRD, completely unrelated to the existing CRD. So conversion webhooks are not applicable in that case.